File: Bots/DotNet/Skills/CodeFirst/WaterfallSkillBot/Dialogs/MessageWithAttachment/MessageWithAttachmentDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.IO;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Builder.Dialogs.Choices;
using Microsoft.Bot.Connector;
using Microsoft.Bot.Schema;

namespace Microsoft.BotFrameworkFunctionalTests.WaterfallSkillBot.Dialogs.Attachments
{
    public class MessageWithAttachmentDialog : ComponentDialog
    {
        private const string Picture = "architecture-resize.png";

        public MessageWithAttachmentDialog()
             : base(nameof(MessageWithAttachmentDialog))
        {
            AddDialog(new ChoicePrompt("AttachmentTypePrompt"));
            AddDialog(new ChoicePrompt("EndPrompt"));
            AddDialog(new WaterfallDialog(nameof(WaterfallDialog), new WaterfallStep[] { SelectAttachmentAsync, HandleAttachmentAsync, FinalStepAsync }));
            InitialDialogId = nameof(WaterfallDialog);
        }

        private static Attachment GetInlineAttachment()
        {
            var imagePath = Path.Combine(Environment.CurrentDirectory, "Dialogs", "MessageWithAttachment", "Files", Picture);
            var imageData = Convert.ToBase64String(File.ReadAllBytes(imagePath));

            return new Attachment
            {
                Name = @$"Files\{Picture}",
                ContentType = "image/png",
                ContentUrl = $"data:image/png;base64,{imageData}",
            };
        }

        private static Attachment GetInternetAttachment()
        {
            // ContentUrl must be HTTPS.
            return new Attachment
            {
                Name = @$"Files\{Picture}",
                ContentType = "image/png",
                ContentUrl = "https://docs.microsoft.com/en-us/bot-framework/media/how-it-works/architecture-resize.png",
            };
        }

        private static async Task<Attachment> GetUploadedAttachmentAsync(WaterfallStepContext stepContext, string serviceUrl, string conversationId, CancellationToken cancellationToken)
        {
            if (string.IsNullOrWhiteSpace(serviceUrl))
            {
                throw new ArgumentNullException(nameof(serviceUrl));
            }

            if (string.IsNullOrWhiteSpace(conversationId))
            {
                throw new ArgumentNullException(nameof(conversationId));
            }

            var imagePath = Path.Combine(Environment.CurrentDirectory, "Files", Picture);
            
            var connector = stepContext.Context.TurnState.Get<IConnectorClient>() as ConnectorClient;
            var attachments = new Bot.Connector.Attachments(connector);
            var response = await attachments.Client.Conversations.UploadAttachmentAsync(
                conversationId,
                new AttachmentData
                {
                    Name = $@"Files\{Picture}",
                    OriginalBase64 = await File.ReadAllBytesAsync(imagePath, cancellationToken),
                    Type = "image/png",
                },
                cancellationToken);

            var attachmentUri = attachments.GetAttachmentUri(response.Id);

            return new Attachment
            {
                Name = $@"Files\{Picture}",
                ContentType = "image/png",
                ContentUrl = attachmentUri,
            };
        }

        private async Task<DialogTurnResult> SelectAttachmentAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            // Create the PromptOptions from the skill configuration which contain the list of configured skills.
            var messageText = "What card do you want?";
            var repromptMessageText = "That was not a valid choice, please select a valid card type.";
            var options = new PromptOptions
            {
                Prompt = MessageFactory.Text(messageText, messageText, InputHints.ExpectingInput),
                RetryPrompt = MessageFactory.Text(repromptMessageText, repromptMessageText, InputHints.ExpectingInput),
                Choices = new List<Choice> { new Choice("Inline"), new Choice("Internet") } 
                
                // This is currently excluded since Attachments endpoint isn't currently implemented in the ChannelServiceHandler
                // new Choice("Upload")} 
            };

            // Ask the user to enter their name.
            return await stepContext.PromptAsync("AttachmentTypePrompt", options, cancellationToken);
        }

        private async Task<DialogTurnResult> HandleAttachmentAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var card = stepContext.Context.Activity.Text;
            var reply = MessageFactory.Text(string.Empty);

            switch (card)
            {
                case "Inline":
                    reply.Text = "This is an inline attachment.";
                    reply.Attachments = new List<Attachment>() { GetInlineAttachment() };
                    break;

                case "Internet":
                    reply.Text = "This is an attachment from a HTTP URL.";
                    reply.Attachments = new List<Attachment>() { GetInternetAttachment() };
                    break;

                case "Upload":
                    break;
                    
                    // Commenting this out since the Attachments endpoint isn't currently implemented in the ChannelService Handler
                    //reply.Text = "This is an uploaded attachment.";

                    //// Get the uploaded attachment.
                    //var uploadedAttachment = await GetUploadedAttachmentAsync(stepContext, stepContext.Context.Activity.ServiceUrl, stepContext.Context.Activity.Conversation.Id, cancellationToken);
                    //reply.Attachments = new List<Attachment>() { uploadedAttachment };

                default:
                    reply.Text = "Invalid choice";
                    break;
            }

            await stepContext.Context.SendActivityAsync(reply, cancellationToken);

            var messageText = "Do you want another type of attachment?";
            var repromptMessageText = "That's an invalid choice.";

            var options = new PromptOptions
            {
                Prompt = MessageFactory.Text(messageText, messageText, InputHints.ExpectingInput),
                RetryPrompt = MessageFactory.Text(repromptMessageText, repromptMessageText, InputHints.ExpectingInput),
                Choices = new List<Choice> { new Choice("Yes"), new Choice("No") }
            };

            return await stepContext.PromptAsync("EndPrompt", options, cancellationToken);
        }

        private async Task<DialogTurnResult> FinalStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var selectedSkillId = ((FoundChoice)stepContext.Result).Value.ToLower();

            if (selectedSkillId == "yes")
            {
                return await stepContext.ReplaceDialogAsync(InitialDialogId, "What attachment type do you want?", cancellationToken);
            }

            return new DialogTurnResult(DialogTurnStatus.Complete);
        }
    }
}

File: Bots/JavaScript/Skills/CodeFirst/WaterfallSkillBot/dialogs/messageWithAttachment/messageWithAttachmentDialog.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { MessageFactory, InputHints } = require('botbuilder');
const { ComponentDialog, ChoicePrompt, WaterfallDialog, ChoiceFactory, DialogTurnStatus } = require('botbuilder-dialogs');
const fs = require('fs');
const path = require('path');

const ATTACHMENT_TYPE_PROMPT = 'AttachmentTypePrompt';
const END_PROMPT = 'EndPrompt';
const WATERFALL_DIALOG = 'WaterfallDialog';

class MessageWithAttachmentDialog extends ComponentDialog {
    /**
     * @param {string} dialogId
     */
    constructor(dialogId) {
        super(dialogId);

        this.picture = 'architecture-resize.png';

        this.addDialog(new ChoicePrompt(ATTACHMENT_TYPE_PROMPT))
            .addDialog(new ChoicePrompt(END_PROMPT))
            .addDialog(new WaterfallDialog(WATERFALL_DIALOG, [
                this.selectAttachment.bind(this),
                this.handleAttachment.bind(this),
                this.finalStep.bind(this)
            ]));

        this.initialDialogId = WATERFALL_DIALOG;
    }

    /**
     * @param {import('botbuilder-dialogs').WaterfallStepContext} stepContext
     */
    async selectAttachment(stepContext) {
        // Create the PromptOptions from the skill configuration which contain the list of configured skills.
        const messageText = 'What card do you want?';
        const repromptMessageText = 'That was not a valid choice, please select a valid card type.';
        const options = {
            prompt: MessageFactory.text(messageText, messageText, InputHints.ExpectingInput),
            retryPrompt: MessageFactory.text(repromptMessageText, repromptMessageText, InputHints.ExpectingInput),
            choices: ChoiceFactory.toChoices(['Inline', 'Internet'])

            // This is currently excluded since Attachments endpoint isn't currently implemented in the ChannelServiceHandler
            // choices: ['Upload']
        };

        return stepContext.prompt(ATTACHMENT_TYPE_PROMPT, options);
    }

    /**
     * @param {import('botbuilder-dialogs').WaterfallStepContext} stepContext
     */
    async handleAttachment(stepContext) {
        const card = stepContext.context.activity.text;
        const reply = MessageFactory.text('');

        switch (card) {
        case 'Inline':
            reply.text = 'This is an inline attachment.';
            reply.attachments = [this.getInlineAttachment()];
            break;

        case 'Internet':
            reply.text = 'This is an attachment from a HTTP URL.';
            reply.attachments = [this.getInternetAttachment()];
            break;

        case 'Upload':
            // Commenting this out since the Attachments endpoint isn't currently implemented in the ChannelService Handler

            // reply.text = 'This is an uploaded attachment.';
            // Get the uploaded attachment.
            // var uploadedAttachment = await this.getUploadedAttachment(stepContext, stepContext.context.activity.serviceUrl, stepContext.context.activity.conversation.id);
            // reply.attachments = [uploadedAttachment];
            break;

        default:
            reply.text = 'Invalid choice';
            break;
        }

        await stepContext.context.sendActivity(reply);

        const messageText = 'Do you want another type of attachment?';
        const repromptMessageText = "That's an invalid choice.";

        return stepContext.prompt(END_PROMPT, {
            prompt: MessageFactory.text(messageText, messageText, InputHints.ExpectingInput),
            retryPrompt: MessageFactory.text(repromptMessageText, repromptMessageText, InputHints.ExpectingInput),
            choices: ChoiceFactory.toChoices(['Yes', 'No'])
        });
    }

    /**
     * @param {import('botbuilder-dialogs').WaterfallStepContext} stepContext
     */
    async finalStep(stepContext) {
        const selectedChoice = stepContext.result.value.toLowerCase();

        if (selectedChoice === 'yes') {
            return stepContext.replaceDialog(this.initialDialogId);
        }

        return { status: DialogTurnStatus.complete };
    }

    /**
     * Returns an inline attachment.
     * @returns {import('botbuilder').Attachment}
     */
    getInlineAttachment() {
        const filepath = path.resolve(__dirname, 'files', this.picture);
        const file = fs.readFileSync(filepath, 'base64');
        return {
            name: `Files\\${ this.picture }`,
            contentType: 'image/png',
            contentUrl: `data:image/png;base64,${ file }`
        };
    }

    /**
     * Returns an attachment to be sent to the user from a HTTPS URL.
     * @returns {import('botbuilder').Attachment}
     */
    getInternetAttachment() {
        return {
            name: `Files\\${ this.picture }`,
            contentType: 'image/png',
            contentUrl: 'https://docs.microsoft.com/en-us/bot-framework/media/how-it-works/architecture-resize.png'
        };
    }

    /**
     * Returns an attachment that has been uploaded to the channel's blob storage.
     * @param {import('botbuilder-dialogs').WaterfallStepContext} stepContext
     * @param {string} serviceUrl
     * @param {string} conversationId
     * @returns {import('botbuilder').Attachment}
     */
    async getUploadedAttachment(stepContext, serviceUrl, conversationId) {
        if (!serviceUrl) {
            throw new Error("Argument 'serviceUrl' is null or empty.");
        }

        if (!conversationId) {
            throw new Error("Argument 'conversationId' is null or empty.");
        }

        const imagePath = path.resolve(__dirname, 'files', this.picture);
        const image = fs.readFileSync(imagePath);

        const connector = stepContext.context.turnState.get(stepContext.context.adapter.ConnectorClientKey);
        const response = await connector.conversations.uploadAttachment(conversationId, {
            name: `Files\\${ this.picture }`,
            type: 'image/png',
            originalBase64: image
        });

        const baseUri = connector.baseUri;
        const attachmentUri = baseUri + (baseUri.endsWith('/') ? '' : '/') + `v3/attachments/${ encodeURI(response.id) }/views/original`;

        return {
            name: `Files\\${ this.picture }`,
            contentType: 'image/png',
            contentUrl: attachmentUri
        };
    }
}

module.exports.MessageWithAttachmentDialog = MessageWithAttachmentDialog;
